File: homework3/saad_amin/q11_vacationDestination.py
#Time Complexity: O(E log V)
#Space : O(V + E)
#Technique: Heaps and BFS(Dijktsra)

import heapq
from collections import defaultdict

def vacationDestinations(edges, origin, k):
    graph = defaultdict(list)

    for city1, city2, time in edges:
        graph[city1].append((city2, time))
        graph[city2].append((city1, time))

    heap = [(0, origin)]
    shortest = {origin: 0}

    while heap:
        curr_time, city = heapq.heappop(heap)

        if curr_time > shortest[city]:
            continue

        for nei, travel_time in graph[city]:
            extra_stop_time = 0 if city == origin else 1
            new_time = curr_time + travel_time + extra_stop_time

            if new_time < shortest.get(nei, float("inf")):
                shortest[nei] = new_time
                heapq.heappush(heap, (new_time, nei))

    count = 0
    for city in shortest:
        if city != origin and shortest[city] <= k:
            count += 1

    return count

edges = [
    ("Boston", "New York", 4),
    ("New York", "Philadelphia", 2),
    ("Boston", "Newport", 1.5),
    ("Washington, D.C.", "Harper's Ferry", 1),
    ("Boston", "Portland", 2.5),
    ("Philadelphia", "Washington, D.C.", 2.5)
]

print(vacationDestinations(edges, "New York", 5))  # 2
print(vacationDestinations(edges, "New York", 7))  # 4
print(vacationDestinations(edges, "New York", 8))  # 6

#time: 40 min

File: Part2Questions/q11_VacationDestination.java
package Part2Questions;

import java.util.*;

public class q11_VacationDestination {
    /**
     * Algorithm used: Dijkstras Dsa used Graph
     * 
     * so my plan is to think of cities as graph wehre each connectoin has a travel time meaning weighted greaph
     * I will start from origin and explore other cities while choosing the path that takes the least time so far.
     * i will keep track of the least time so far. then count how many cities i can reach within the time limit given
     */

    static class Edge {
        String city;
        double time;

        Edge(String city, double time) {
            this.city = city;
            this.time = time;
        }
    }

    static class State {
        String city;
        double totalTime;

        State(String city, double totalTime) {
            this.city = city;
            this.totalTime = totalTime;
        }
    }

    public static int vacationDestination(List<String[]> routes, String origin, double timeLimit) {
        Map<String, List<Edge>> graph = new HashMap<>();

        
        for (String[] route : routes) {
            String from = route[0];
            String to = route[1];
            double time = Double.parseDouble(route[2]);

            graph.putIfAbsent(from, new ArrayList<>());
            graph.putIfAbsent(to, new ArrayList<>());

            graph.get(from).add(new Edge(to, time));
            graph.get(to).add(new Edge(from, time));
        }

        if (!graph.containsKey(origin)) {
            return 0;
        }

        Map<String, Double> shortestTime = new HashMap<>();

        PriorityQueue<State> pq = new PriorityQueue<>(
            (a, b) -> Double.compare(a.totalTime, b.totalTime)
        );

        shortestTime.put(origin, 0.0);
        pq.add(new State(origin, 0.0));

        while (!pq.isEmpty()) {
            State current = pq.poll();

            if (current.totalTime > shortestTime.get(current.city)) {
                continue;
            }

            for (Edge edge : graph.get(current.city)) {
                double newTime = current.totalTime + edge.time;

                // Add stopover cost only when current city is not the origin
                if (!current.city.equals(origin)) {
                    newTime += 1;
                }

                if (!shortestTime.containsKey(edge.city) || newTime < shortestTime.get(edge.city)) {
                    shortestTime.put(edge.city, newTime);
                    pq.add(new State(edge.city, newTime));
                }
            }
        }

        int count = 0;

        for (String city : shortestTime.keySet()) {
            if (!city.equals(origin) && shortestTime.get(city) <= timeLimit) {
                count++;
            }
        }

        return count;
    }

    public static void main(String[] args) {
        List<String[]> routes = new ArrayList<>();

        routes.add(new String[]{"Boston", "New York", "4"});
        routes.add(new String[]{"New York", "Philadelphia", "2"});
        routes.add(new String[]{"Boston", "Newport", "1.5"});
        routes.add(new String[]{"Washington, D.C.", "Harper's Ferry", "1"});
        routes.add(new String[]{"Boston", "Portland", "2.5"});
        routes.add(new String[]{"Philadelphia", "Washington, D.C.", "2.5"});

        System.out.println(vacationDestination(routes, "New York", 5));  // 2
        System.out.println(vacationDestination(routes, "New York", 7));  // 4
        System.out.println(vacationDestination(routes, "New York", 8));  // 6
    }
}
// Time is O(M log N) space is O(N+M)
//  N = number of cities M = number of routes
